mindocr/data/transforms/general_transforms.py

```diff
@@ -205,48 +210,47 @@ def __call__(self, data: dict):
             image
             (polys)
         """
-        scale = np.random.uniform(*self._range)
-        data['image'] = cv2.resize(data['image'], dsize=None, fx=scale, fy=scale)
+        if random.random() < self._p:
+            scale = np.random.uniform(*self._range)
+            data['image'] = cv2.resize(data['image'], dsize=None, fx=scale, fy=scale)
 
-        if 'polys' in data:
-            data['polys'] *= scale
+            if 'polys' in data:
+                data['polys'] *= scale
         return data
 
 
 class RandomCropWithBBox:
     """
-    Randomly cuts a crop from an image along with polygons.
+    Randomly cuts a crop from an image along with polygons in the way that the crop doesn't intersect any polygons
+    (i.e. any given polygon is either fully inside or fully outside the crop).
 
     Args:
-        max_tries: number of attempts to try to cut a crop with a polygon in it.
+        max_tries: number of attempts to try to cut a crop with a polygon in it. If fails, scales the whole image to
+                   match the `crop_size`.
         min_crop_ratio: minimum size of a crop in respect to an input image size.
         crop_size: target size of the crop (resized and padded, if needed), preserves sides ratio.
+        p: probability of the augmentation being applied to an image.
     """
-    def __init__(self, max_tries=10, min_crop_ratio=0.1, crop_size=(640, 640)):
+    def __init__(self, max_tries=10, min_crop_ratio=0.1, crop_size=(640, 640), p: float = 0.5):
         self._crop_size = crop_size
         self._ratio = min_crop_ratio
         self._max_tries = max_tries
+        self._p = p
 
     def __call__(self, data):
-        start, end = self._find_crop(data)
+        if random.random() < self._p:   # cut a crop
+            start, end = self._find_crop(data)
+        else:                           # scale and pad the whole image
+            start, end = np.array([0, 0]), np.array(data['image'].shape[:2])
+
         scale = min(self._crop_size / (end - start))
 
         data['image'] = cv2.resize(data['image'][start[0]: end[0], start[1]: end[1]], None, fx=scale, fy=scale)
+        data['actual_size'] = np.array(data['image'].shape[:2])
         data['image'] = np.pad(data['image'],
                                (*tuple((0, cs - ds) for cs, ds in zip(self._crop_size, data['image'].shape[:2])), (0, 0)))
 
-        start, end = start[::-1], end[::-1]     # convert to x, y coord
-        new_polys, new_texts, new_ignores = [], [], []
-        for _id in range(len(data['polys'])):
-            # if the polygon is within the crop
-            if (data['polys'][_id].max(axis=0) > start).all() and (data['polys'][_id].min(axis=0) < end).all():   # NOQA
-                new_polys.append((data['polys'][_id] - start) * scale)
```

Comment:
Is there a strong reason to move data['texts'] and data['ignore_tags'] out of RandomCropWithBBox? If the user forgets to use ValidatePolygons after RandomCropWithBBox, wouldn't this lead to some errors?

Reply:
For the reason you mentioned below, `rotation` also causes some polygons to be outside of the visible region. Moreover, for our current pipeline, polygon validation is applied in cropping and binary map generation, but it also needed in the rotation step. So, to merge all 3 validation steps together, I introduced the `ValidatePolygons` class.